File: pt-1/3918.sum-of-primes-between-number-and-its-reverse.go
package leetcode

//
// @lc app=leetcode id=3918 lang=golang
//
// [3918] Sum of Primes Between Number and Its Reverse
//
// Notes
// Reverse n, sieve all primes up to max(n, reverse(n)), and build a prefix sum
// of prime values. The range answer is then prefix[right] - prefix[left-1].
// Time: O(R log log R). Space: O(R), where R is max(n, reverse(n)).
//
// @lc code=start

func SumOfPrimesInRange3918(n int) int {
	reversedN := reverseInt3918(n)
	left, right := n, reversedN
	if left > right {
		left, right = right, left
	}

	isPrime := sieve3918(right)
	prefix := make([]int, right+1)
	for value := 1; value <= right; value++ {
		prefix[value] = prefix[value-1]
		if isPrime[value] {
			prefix[value] += value
		}
	}
	if left > 0 {
		return prefix[right] - prefix[left-1]
	}
	return prefix[right]
}

func reverseInt3918(n int) int {
	reversed := 0
	for n > 0 {
		reversed = reversed*10 + n%10
		n /= 10
	}
	return reversed
}

func sieve3918(limit int) []bool {
	isPrime := make([]bool, limit+1)
	if limit < 2 {
		return isPrime
	}
	for i := 2; i <= limit; i++ {
		isPrime[i] = true
	}
	for p := 2; p*p <= limit; p++ {
		if isPrime[p] {
			for multiple := p * p; multiple <= limit; multiple += p {
				isPrime[multiple] = false
			}
		}
	}
	return isPrime
}

// @lc code=end

File: pt-1/3918.sum-of-primes-between-number-and-its-reverse.java
/*
 * @lc app=leetcode id=3918 lang=java
 *
 * [3918] Sum of Primes Between Number and Its Reverse
 *
 * Reverse n, sieve primes up to max(n, reverse), build prefix sums of prime
 * values, and subtract the range before min(n, reverse).
 *
 * Time: O(R log log R). Space: O(R), R = max(n, reverse(n)).
 */

// @lc code=start
class Solution {
    public long sumOfPrimesInRange(int n) {
        int reversed = Integer.parseInt(new StringBuilder(Integer.toString(n)).reverse().toString());
        int left = Math.min(n, reversed);
        int right = Math.max(n, reversed);
        boolean[] isPrime = sieve(right);
        long[] prefix = new long[right + 1];
        for (int value = 1; value <= right; value++) {
            prefix[value] = prefix[value - 1] + (isPrime[value] ? value : 0);
        }
        return prefix[right] - (left > 0 ? prefix[left - 1] : 0);
    }

    private boolean[] sieve(int limit) {
        boolean[] isPrime = new boolean[limit + 1];
        if (limit < 2) {
            return isPrime;
        }
        for (int i = 2; i <= limit; i++) {
            isPrime[i] = true;
        }
        for (int p = 2; p * p <= limit; p++) {
            if (isPrime[p]) {
                for (int multiple = p * p; multiple <= limit; multiple += p) {
                    isPrime[multiple] = false;
                }
            }
        }
        return isPrime;
    }
}
// @lc code=end
